### DISCUSSION TITLE
Interval with MissedTickBehavior::Delay waits less than period

### DISCUSSION THREAD

**Discussion by nearwall** [tokio-rs/tokio#5620] [Q&A]
Hello guys!

I created a test task with constant iteration delay and tried to manage stopped time with `tokio::time::advance`.
For 100ms delay I got 200ms in output. A bit unexpectedly, but ok.
And I tried to fix it with the context switch before `tokio::time::advance(100ms)` with `yield_now` alternative and got nice
100ms results.
But after that I set 7ms delay and got something unexpected: **98ms** delay for `MissedTickBehavior::Delay`:

```
use crossbeam_queue::ArrayQueue;
use std::sync::Arc;
use tokio::time::MissedTickBehavior;

// 1 attempt
// const TIME_STEP: u64 = 100;

// 2 attempt
const TIME_STEP: u64 = 7;


#[tokio::test(flavor = "current_thread", start_paused = true)]
async fn interval_test() {
    let timestamps = Arc::new(ArrayQueue::new(32));

    let timestamps_ms = timestamps.clone();
    tokio::task::spawn(async move {
        let mut intrvl = tokio::time::interval(tokio::time::Duration::from_millis(100));
        intrvl.set_missed_tick_behavior(MissedTickBehavior::Delay);

        intrvl.tick().await;

        for _ in 0..10 {
            let instant = tokio::time::Instant::now();
            intrvl.tick().await;

            timestamps_ms
                .push(instant.elapsed())
                .unwrap();
        }
    });

    // just wait the task
    for _ in 0..200 {
        // `yield_now` alternative
		tokio::time::advance(tokio::time::Duration::from_nanos(0)).await;

        tokio::time::advance(tokio::time::Duration::from_millis(TIME_STEP)).await; 
    }

    while let Some(time_ms) = timestamps.pop() {
        println!("Time delta: {time_ms:?}");
    }
}
```

Test output:

```
Time delta: 105ms
Time delta: 98ms
Time delta: 98ms
Time delta: 105ms
Time delta: 105ms
Time delta: 98ms
Time delta: 98ms
Time delta: 105ms
Time delta: 105ms
Time delta: 98ms

```

-  Ubuntu 22.10
- rustc 1.69.0-nightly (dc1d9d50f 2023-01-31)
- tokio = { version = "1.27.0", features = ["full", "test-util"] }

I thought, that `MissedTickBehavior::Delay` always has equal or bigger ticks than the period...

Maybe someone has any ideas about such strange behaviour?

**Comment by nearwall** [comment]
It seems, that there could be my problem: https://docs.rs/tokio/latest/src/tokio/time/interval.rs.html#478
We have such check for missed tick:
`let next = if now > timeout + Duration::from_millis(5)`
And in my case the first delay is 105ms = period + 5ms.  105 unfortunately is in allowed range, and I don't get the missed tick condition. The next delay is 98ms, but for Interval it is 98 + 5ms

**Reply by Darksonn** [reply]
This is expected behavior.
